docs_src/design/adr/device-service/0020-invoking-functions.md

```diff
@@ -0,0 +1,74 @@
+# Invoking Device Functions
+ 
+
+## Status
+
+**Draft**
+
+## Context
+This ADR presents a mechanism for invoking functions on a device.
+
+## Existing Behavior
+
+Device access in EdgeX is focussed on 'Readings' and 'Settings'. A device may
+present values which are readable, representing some measurement or status,
+such as a temperature or vibration reading. Also a setting such as a the
+required speed of a fan motor may be written to and read back.
+
+What is not supported directly is 'Functions', where a device may be commanded
+to do something. Examples could include perform a self-test, go into standby
+mode for an hour, invalidate an access key.
+
+While such operations may modelled using virtual resources in a device profile,
+this is unintuitive.
+
+## Decision
+
+1. Add a new section to device profiles describing functions
+
+{
+  "deviceFunctions":
+  [
+    {
+      "name": "Name by which the function is accessed",
+      "description": "Readable description of the function",
+      "attributes": { device-specific parameters which select this function },
+      "parameters":
+      {
+        "in":
+        [
+          {
+            "name": "Parameter name",
+            "description": "description of what the parameter controls",
+            "type": "Any of the usual EdgeX data types"
```

Comment:
Here are two fields on device function parameters to consider:
- defaultValue (optional): Contains a default value if the parameter isn't supplied.
- range (optional): min/max values for numeric types.

Reply:
Added these (but used maximum / minimum for consistency with device resources)

Reply:
Min/max makes sense.  Thanks.